### ISSUE TITLE
RFC: Indexing API

### ISSUE THREAD

**Issue by aboodman** [rocicorp/repc#201]
# Background

Replicache seeks to enable "near-instant" reads for client-side applications. We define "near-instant" as <20ms for "a few screenfuls of data" for common applications.

20ms is an important threshold.  It's too slow to *animate* the results of Replicache queries, but fast enough to be below the perception threshold for actions like tapping a button and seeing results.

This means that practice developers can often treat Replicache as local, and don't need to worry about progress UI, concurrency, and other complexities that usually come with talking to distant data sources.

Initial results indicate that the maximum throughput we can currently get out of IndexedDB (with JSON payloads) is on order of 50MB/s. There are various improvements coming to wasm and wasm/rust bindings that will improve this, but given that our max cache size is currently 20MB, and we want to support much larger caches in the future this is a non-starter. At 50MB/s it takes 400ms to scan a 20MB cache.

# Indexes

In order to eliminate the need to constantly scan the entire cache, indexes have long been a desired feature of Replicache. This bug proposes a simple API to define and use indexes in Replicache, and an initial implementation of that API.

# Prior Art

- IndexedDB: https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/createIndex
- CouchDB: https://docs.couchdb.org/en/stable/ddocs/views/intro.html#what-is-a-view
- RethinkDB: https://rethinkdb.com/docs/secondary-indexes/python/
- MongoDB: https://docs.mongodb.com/manual/indexes/

# API (expressed in TypeScript)

```
// Creates an index on the subset of the keyspace with the specified prefix.
// The index is over the result of evaluating the given [json path](https://goessner.net/articles/JsonPath/).
//
// The result of evaluating the indexPath can be:
// - a single boolean, string, or number value: The value is encoded lexicographically as the index key.
//.   sort order amongst different json types is not guaranteed and should not be relied upon
// - an array of boolean, string, or number values: Each value is encoded as above and a separate index entry is
//    added for each.
// - any other type, or undefined: No index entry is generated.
//
// Index creation is asynchronous and may take several seconds. If the page is unloaded while index creation is
// happening, it will *not* automatically resume on next load. If an index is created with different parameters
// than it previously had, that is treated as a silent drop and re-create. If an index is created with identical
// parameters, then it will be an instant no-op.
async function createIndex(name: string, indexPath: string, keyspacePrefix?: string = ''): Promise<void>;

// Drops the specified index if it exists.
async function dropIndex(name: string): Promise<void>;
```

We also update the existing `scan()` method to include a new `indexName` field:

```
// If indexName is specified then the prefix and start values apply to the range of the specified index to scan.
//
// ScanResult always returns key/value pairs. The key is either the primary key, or the secondary key from the index
// (when indexName is used). The value is always a primary value (Replicache maps back to index values internally).
scan(__namedParameters?: { prefix: string; start: undefined | ScanBound, indexName: string | undefined }): ScanResult
```

# Implementation

## Data Storage

Index definitions are stored in a separate part of the kv store's key space from other data. Definitions are in flatbuffer format.

Indexes themselves are stored as separate `prolly::Map` values from the "primary" map. The benefit of doing this is that drops are fast and easy. We add a new `indexes` field to `Commit` that is a map of indexName to root hash for the corresponding `prolly::Map`.

## Maintenance

Index maintenance is initially synchronous. It is possible to do clever asynchronous index maintenance in the future but that's out of scope for now.

In psuedocode, the following would run for each put/del:

```
- for each relevant index:
  - compute old indexed values:
    - remove index entries
  - compute new indexed values:
    - add index entries
```

## Scan Changes

I don't think the scan changes are significant. Basically the same code can be used except over a different `prolly::Map` and with the addition of then looking up the relevant items in the primary map.

## Open Questions

* How do multiple tabs interact with this API. Two different tabs could be on two different versions of the application and require two different indexes? If they both use the same index name, then they would fight over the definition and I suppose the second tab to load would break the first. If they encode a version in the index name, then we end up with old indexes slowing things down that aren't needed.

**Comment by phritz** [comment]
> Replicache seeks to enable near-instantaneous reads for client-side applications. For sake of discussion we define this as <20ms for "the amount of data a developer needs to render a view".

OK so working with the numbers above assuming 50MB/s scan speed that's 20ms/MB. If scanning the index for "the amount of data a developer needs to render a view" should take 20ms or less does that imply that that amount of data is <= 1 MB?

> This means that developers cannot necessarily animate the results of Replicache queries, but that users are typically not going to be able to notice the delay between e.g., tapping a button and seeing the results on screen.

This is slightly different than what you said above as it includes a write transaction. Maybe this doesn't matter, but asking anyway. Are there two things here? 1. startup: read N bytes of data to render a view in M ms and 2. update: update X keys and fire all subscriptions within Y ms? Or maybe if we can do 1 we can do 2? 

That's sort of a long way of saying that if we are picking a metric to optimize for I want to be sure I understand it.

> Index creation is asynchronous and may take awhile. If the page is unloaded while index creation is
happening, it will *not* automatically resume on next load. 

Does this mean that when they update their app (reload the page say and get new code) and want to add an index they have to keep code around that checks on page load whether it was actually created and to kick it off if not? Can they distinguish between 'hasn't yet been created need to kick it off' vs 'tried to create it and hit an error so couldn't'? I might have expected that if they ask for the index to be created we will keep trying until we succeed or fail, that is, that if creation is interrupted by navigating away then we'd resume probably from scratch next time.

Does it matter to their app whether they hear about failure to create an index synchronously vs some asynchronous mechanism (maybe we have a subscription for that?).

Other qs:
- numerical scans with an index: they happen in lexicographic or numerical order?
- indexes in the kv store would get a key prefix like '/i/' yes? i am going to want to add something like '/m/' for metadata the customer can use to store cookies.
- good point about versioning. i think it needs to be up to the customer to clean up old indexes. we can of course give hints if want about unused indexes.

**Comment by aboodman** [comment]
On Tue, Sep 29, 2020 at 12:59 PM Phritz <notifications@github.com> wrote:

> Replicache seeks to enable near-instantaneous reads for client-side
> applications. For sake of discussion we define this as <20ms for "the
> amount of data a developer needs to render a view".
>
> OK so working with the numbers above assuming 50MB/s scan speed that's
> 20ms/MB. If scanning the index for "the amount of data a developer needs to
> render a view" should take 20ms or less does that imply that that amount of
> data is <= 1 MB?
>
I think so, which checks out for me intuitively.

> This means that developers cannot necessarily animate the results of
> Replicache queries, but that users are typically not going to be able to
> notice the delay between e.g., tapping a button and seeing the results on
> screen.
>
> This is slightly different than what you said above as it includes a write
> transaction. Maybe this doesn't matter, but asking anyway. Are there two
> things here? 1. startup: read N bytes of data to render a view in M ms and
> 2. update: update X keys and fire all subscriptions within Y ms? Or maybe
> if we can do 1 we can do 2?
>
Good point. Yes, I'm concerned about write/read latency too. I think we
should have a goal to service common write/read txs below the
tap-perception threshold. I remember having a mental bookmark that this
number was ~20ms, but I failed to find research backing that up.

> That's sort of a long way of saying that if we are picking a metric to
> optimize for I want to be sure I understand it.
>
> Index creation is asynchronous and may take awhile. If the page is
> unloaded while index creation is
> happening, it will *not* automatically resume on next load.
>
> Does this mean that when they update their app (reload the page say and
> get new code) and want to add an index they have to keep code around that
> checks on page load whether it was actually created and to kick it off if
> not?
>
I think so. The alternative is that we promise to complete index creation
... eventually ... but in that case they still need to have code that
checks for index presence and creates it on first run anyway, so I didn't
see the difference in practice.

> Can they distinguish between 'hasn't yet been created need to kick it off'
> vs 'tried to create it and hit an error so couldn't'?
>
I was going to say yes, but actually ... multiple tabs. I'll think about
this more.

> I might have expected that if they ask for the index to be created we will
> keep trying until we succeed or fail, that is, that if creation is
> interrupted by navigating away then we'd resume probably from scratch next
> time.
>
Yes I thought maybe *not* doing that would be "easier" for some combination
of us/customer, but not clear now.

> Does it matter to their app whether they hear about failure to create an
> index synchronously vs some asynchronous mechanism (maybe we have a
> subscription for that?).
>
I think it would basically be an internal error (e.g. our fault) if index
creation fails right? I can't think of any legitimate reason for that to
occur. If so I think just putting it into console would be fine.

**Comment by aboodman** [comment]
I feel like a really good benchmark for us to monitor and work to improve
would be:

1. Define a number for tap perception threshold
2. Work to increase the amount of data we can write/read under that
threshold

**Comment by phritz** [comment]
FYI indexing strings: https://en.wikipedia.org/wiki/Unicode_equivalence

**Comment by aboodman** [comment]
There's one last sad challenge here: multiple tabs.

If we do nothing then I think what I've outlined above will be *correct* it will just do extra wasteful work in one of the tabs.

- If you have two tabs, both with version A of the app open.
- Then both update around the same time (perhaps they auto reload) to version B.
- They both get asked for index Q.
- They both see that index Q doesn't exist and begin making it (since they do this in a read tx, they proceed in parallel)
- tab 1 gets his write tx and lands index Q first.
- tab 1 then begins maintaining index Q as normal. If a write happens, it updates the index synchronously.
- tab 2 gets his write tx and goes to land index Q.

We need to be careful when we go to land an index to first check that it hasn't already been landed by somebody else. Otherwise this all works.

**Comment by aboodman** [comment]
OK thinking this through, I feel it's not worth the complexity right now to try and do anything other than synchronous index maintenance.

By "synchronous" I mean that whenever we want to do something to an index -- either create it or update it, we do it immediately, inside a write tx. So like:

- create index: open a write tx, read entire cache, create index, commit
- update index: inside parent write tx, calculate index changes, commit

Multiple tabs greatly complicate any other strategy so I think that if we eventually decide we want to do something fancier for either creation or maintenance, we will first have to implement leader election across tabs so that we don't have any risk of multiple tabs trying to do same thing.

Here is my concrete proposal:

# Schema changes

```
table IndexDefinition {
  name: string;
  key_prefix: string;
  index_path: string;
}

table Index {
  definition: IndexDefinition;
  value_hash: string;
}

// Note: an important thing I realized is that indexes *have* to go under commit because a
// sync fork requires its own forks of the indexes too!
table Commit {
  ...
  // Invariant (checked at validate time) that indexes are unique by name.
  // Question: should we also sort them?
  indexes: [IndexDefinition];
}
```

# Code changes

## Load/Save code for FBS

Nothing to say here, should be straightforward

## Index Creation

Implement `db::Write::create_index`
- Validate index doesn't already exist and other invariants of spec
- Find the lightest-weight possible way to parse a subset out a subset of a json value in Rust
- Generalize prolly load/save code
- Iterate keys with specified prefix and create writes for index
  - Thinking currently that indexes will be "covered": https://stackoverflow.com/questions/62137/what-is-a-covered-index
- Flush all prolly trees during db::Write::commit, not just the main one

Implement `embed::do_create_index`
- Needs to create a `db::Write`, call `create_index`, then `commit()`
- Wire up and test

## Index Scan

- If we use covered indexes, change to `db::Read::scan()` is just passing in a different map to `scan` helper

## Index Maintenance

- Add to prolly::Map ability to extract changed keys
- Iterate changed keys to get those that match prefix
- Implement `update_indices` and call for each updated key
  - Extract code from `db::create_index` for this
- `db::Write::commit` should otherwise work

## Index drop

- Add db::Write::drop_index

## Bindings Changes

- Add `create_index`, `get_index`, and `drop_index`
- Add `indexName` to `scan()`

**Comment by phritz** [comment]
>  I feel it's not worth the complexity right now to try and do anything other than synchronous index maintenance. ... By "synchronous" I mean that whenever we want to do something to an index -- either create it or update it, we do it immediately, inside a write tx

Huge cheer of support from over here.
